File: webapp/src/schemas/tracked-application.ts
import { z } from 'zod';

export const applicationOutcomeSchema = z.enum(['applied', 'interview', 'offer', 'rejected']);
export type ApplicationOutcome = z.infer<typeof applicationOutcomeSchema>;

export const trackedApplicationSchema = z.object({
  id: z.string(),
  profile_id: z.string(),
  job_posting_id: z.string().nullable(),
  role: z.string(),
  company: z.string(),
  location: z.string().nullable(),
  match_score: z.number().nullable(),
  status: applicationOutcomeSchema,
  source: z.string(),
  applied_at: z.string(),
});

export const trackedApplicationListSchema = z.object({
  items: z.array(trackedApplicationSchema),
  total: z.number(),
  counts: z.record(z.string(), z.number()),
});

export type TrackedApplication = z.infer<typeof trackedApplicationSchema>;
export type TrackedApplicationList = z.infer<typeof trackedApplicationListSchema>;

File: backend/app/models/tracked_application.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.models.common import ApiModel

ApplicationOutcomeLiteral = Literal["applied", "interview", "offer", "rejected"]


class TrackedApplicationOutput(ApiModel):
    id: str
    profile_id: str
    job_posting_id: str | None
    role: str
    company: str
    location: str | None
    match_score: float | None
    status: ApplicationOutcomeLiteral
    source: str
    applied_at: datetime


class TrackedApplicationListResponse(ApiModel):
    items: list[TrackedApplicationOutput]
    total: int
    counts: dict[str, int]


class TrackedApplicationCreateRequest(ApiModel):
    profile_id: str
    job_posting_id: str | None = None
    role: str = Field(min_length=1, max_length=500)
    company: str = Field(min_length=1, max_length=300)
    location: str | None = None
    match_score: float | None = Field(default=None, ge=0, le=100)
    status: ApplicationOutcomeLiteral = "applied"
    source: str = Field(default="quick_apply", max_length=40)


class TrackedApplicationUpdateRequest(ApiModel):
    status: ApplicationOutcomeLiteral
